src/formats/SMI.cpp

```diff
@@ -0,0 +1,705 @@
+#include <fmt/ostream.h>
+
+#include "chemfiles/formats/SMI.hpp"
+
+#include "chemfiles/ErrorFmt.hpp"
+#include "chemfiles/File.hpp"
+#include "chemfiles/Frame.hpp"
+#include "chemfiles/periodic_table.hpp"
+#include "chemfiles/warnings.hpp"
+#include "chemfiles/utils.hpp"
+
+using namespace chemfiles;
+
+template<> FormatInfo chemfiles::format_information<SMIFormat>() {
+    return FormatInfo("SMI").with_extension(".smi").description(
+        "SMI text format"
+    );
+}
+
+/// Fast-forward the file for one step, returning `false` if the file does
+/// not contain one more step.
+static bool forward(TextFile& file);
+/// See if all values in array are true
+static bool all(const std::vector<bool>& vec);
+
+SMIFormat::SMIFormat(const std::string& path, File::Mode mode, File::Compression compression)
+    : file_(TextFile::open(std::move(path), mode, compression))
+{
+    while (!file_->eof()) {
+        auto position = file_->tellg();
+        if (!file_ || position == std::streampos(-1)) {
+            throw format_error("IO error while reading '{}' as SMI", path);
+        }
+        if (forward(*file_)) {
+            steps_positions_.push_back(position);
+        }
+    }
+    file_->rewind();
+}
+
+size_t SMIFormat::nsteps() {
+    return steps_positions_.size() - 1;
+}
+
+void SMIFormat::read_step(const size_t step, Frame& frame) {
+    assert(step < steps_positions_.size());
+    file_->seekg(steps_positions_[step]);
+    read(frame);
+}
+
+/// Attempts to read a number from the string. The `start` argument is updated
+/// to the last numberic character in `smiles`.
+static size_t read_number(const std::string& smiles, size_t& start) {
+    size_t number = 0;
+    if (start >= smiles.size() || std::isdigit(smiles[start]) == 0) {
+        --start;
+        return number;
+    }
+    do {
+        number *= 10;
+        number += static_cast<size_t>(smiles[start] - '0');
+        ++start;
+    } while (start < smiles.size() && (std::isdigit(smiles[start]) != 0));
+    --start;
+    return number;
+}
+
+static std::set<char> aromatic_organic {'b', 'c', 'n', 'o', 's', 'p'};
+
+// Note: Including H here is non-standard, but allows for HC(H)(H)H
+static std::set<std::string> aliphatic_organic {"B", "C", "N", "O", "S", "P",
+                                                "F", "Cl", "Br", "I", "H"};
+
+// Valid chirality tags
+static std::set<std::string> chirality_tags {"TH", "SP", "TB", "OH", "AL"};
+
+static std::string read_property_atom(const std::string& smiles, size_t& i) {
+    auto old = i;
+    if (smiles[i] == '\'') {
+        do {
+             ++i;
+        } while( i < smiles.size() && smiles[i] != '\'');
+        ++i;
+        return smiles.substr(old + 1, i - old - 2);
+    } else {
+        do {
+            ++i;
+        } while( i < smiles.size() && std::islower(smiles[i]) != 0);
+        return smiles.substr(old, i - old);
+    }
+}
+
+Atom& SMIFormat::add_atom(Topology& topol, const std::string& atom_name) {
```

Comment:
you can pass atom_name by value and move it to construct the new atom (and move it too at the call site)